### PR TITLE
[RFC] Misc. macro cleanup

### PR DESCRIPTION
**Pull request by ghost**
commit message:

```
- ROOT_UID's comment is misleading, as it's always defined to 0.

- SEEK_{SET,END} & O_NOFOLLOW should already be defined on Unix-like
  systems in <stdio.h> and <fcntl.h>, respectively.  In any case,
  neither of those #ifdef blocks should be in the middle of source files.

- The S_IS{LNK,DIR,...} macros should only be undefined on Windows.
```

### CODE REVIEW

**Comment by sethjackson** [comment]
Huge :+1: to this entire PR. There are also **more** defines of `SEEK_SET` and `SEEK_END` in `src/nvim/misc1.c` that should be removed.

**Comment by Pyrohh** [comment]
> SEEK_SET and SEEK_END in src/nvim/misc1.c that should be removed.

Thanks, I added those too. I'll do a little more searching and see if we can add more crud to this PR.

**Comment by sethjackson** [comment]
> I'll do a little more searching and see if we can add more crud to this PR.

:+1: 

How are you finding these BTW? My grep-fu is weak. :(

**Comment by Pyrohh** [comment]
It's slow, but it works:

```
git grep -h -E '^[ \t]*#[ \t]*if(n?def| ?!?defined)' |
sed -E 's/^[ \t]*#[ \t]*if(n?def| ?!?defined)//' |
sed -E 's/[()]//g' |
awk '{print $1}' |
sort |
uniq
```

**Comment by Pyrohh** [comment]
Forgot to mention:

> My grep-fu is weak. :(

Check out http://regexone.com/, it's a well designed tutorial IMO. Regex is for the most part pretty easy; maybe that's why it's so ugly...

**Comment by sethjackson** [comment]
> Check out http://regexone.com/, it's a well designed tutorial IMO.

:+1: thanks.

I was wondering if you were using clang to find unused symbols or something but I guess it wouldn't really help in this instance since we are using them.

Also just so someone doesn't have to later you might want to add `<uv.h>` to `src/nvim/os/win_defs.h` from equalsraf@956e9eb since we will eventually need that for `S_IFLNK`.

**Comment by Pyrohh** [comment]
> Also just so someone doesn't have to later you might want to add <uv.h> to src/nvim/os/win_defs.h from equalsraf@956e9eb since we will eventually need that for S_IFLNK.

I think that would be better for a separate PR, so, when we actually need it (assuming we don't at the moment).

**Comment by sethjackson** [comment]
> I think that would be better for a separate PR, so, when we actually need it (assuming we don't at the moment).

We don't need it right now because we are only compiling third-party. So :+1: and LGTM.

**Comment by Pyrohh** [comment]
Thanks @sethjackson.

File: src/nvim/vim.h [hunk 1]
```diff
@@ -27,9 +27,6 @@ Error: configure did not run properly.Check auto/config.log.
 # endif
 #endif
 
-/* user ID of root is usually zero, but not for everybody */
-#define ROOT_UID 0
```
Comment (justinmk) [inline src/nvim/vim.h]: Even if the comment is misleading, we should not be using a magic number in the source code...

Comment (sethjackson) [inline src/nvim/vim.h]: I think that callers of `getuid()` would (or should) know that root's UID will always be `0`.

Comment (Pyrohh) [inline src/nvim/vim.h]: > would (or should) know that root's UID will always be 0

This is the reason I made the change; I'm all for avoiding magic numbers, but the UID/GID of root seems too common to necessitate a macro.

Comment (justinmk) [inline src/nvim/vim.h]: That's what everyone says when they use magic numbers. There's no good reason to remove this define--its existence costs nothing and adds some finite amount of clarity.

Comment (Pyrohh) [inline src/nvim/vim.h]: > adds some finite amount of clarity.

I think so too, I'll add it back.

**Pull request merged**